File: hermes/reman-agentic/file_access.py
"""Fail-closed local PDF boundary for the Hermes connector."""

import os
import stat
from pathlib import Path

from .client import RemanError


def _configured_values():
    return [value.strip() for value in os.environ.get("REMAN_AGENT_ALLOWED_PDF_DIRS", "").split(os.pathsep) if value.strip()]


def has_configured_pdf_roots():
    return bool(_configured_values())


def allowed_pdf_roots():
    values = _configured_values()
    if not values:
        raise RemanError("reman_pdf_roots_not_configured")
    roots = []
    for value in values:
        declared = Path(value).expanduser()
        if not declared.is_absolute():
            raise RemanError("reman_pdf_root_invalid")
        try:
            canonical = declared.resolve(strict=True)
        except OSError:
            raise RemanError("reman_pdf_root_invalid") from None
        if not canonical.is_dir():
            raise RemanError("reman_pdf_root_invalid")
        pair = (declared.absolute(), canonical)
        if pair not in roots:
            roots.append(pair)
    return roots


def _relative_to_allowed_root(raw_path, roots):
    path = Path(raw_path).expanduser()
    if not path.is_absolute() or ".." in path.parts:
        raise RemanError("reman_pdf_path_denied")
    lexical = path.absolute()
    for declared, canonical in roots:
        for base in (declared, canonical):
            try:
                return canonical, lexical.relative_to(base)
            except ValueError:
                continue
    raise RemanError("reman_pdf_path_denied")


def _validate_path(root, relative):
    current = root
    parts = relative.parts
    if not parts:
        raise RemanError("reman_pdf_not_regular")
    for index, part in enumerate(parts):
        current = current / part
        try:
            details = os.lstat(current)
        except OSError:
            raise RemanError("reman_pdf_invalid_or_missing") from None
        if stat.S_ISLNK(details.st_mode):
            raise RemanError("reman_pdf_symlink_denied")
        if index < len(parts) - 1 and not stat.S_ISDIR(details.st_mode):
            raise RemanError("reman_pdf_path_denied")
    if not stat.S_ISREG(details.st_mode):
        raise RemanError("reman_pdf_not_regular")
    try:
        if current.resolve(strict=True) != current:
            raise RemanError("reman_pdf_symlink_denied")
    except OSError:
        raise RemanError("reman_pdf_invalid_or_missing") from None
    return current, details


def _same_file(left, right):
    return (
        left.st_dev,
        left.st_ino,
        left.st_size,
        left.st_mtime_ns,
        left.st_ctime_ns,
    ) == (
        right.st_dev,
        right.st_ino,
        right.st_size,
        right.st_mtime_ns,
        right.st_ctime_ns,
    )


def read_allowed_pdf(raw_path, roots, max_bytes=None, before_open=None):
    root, relative = _relative_to_allowed_root(raw_path, roots)
    path, before = _validate_path(root, relative)
    if path.suffix.lower() != ".pdf":
        raise RemanError("reman_pdf_invalid_or_missing")
    if before.st_size <= 0 or (max_bytes is not None and before.st_size > max_bytes):
        raise RemanError("reman_pdf_exceeds_grant_limit")
    if before_open:
        before_open(path)
    flags = os.O_RDONLY | getattr(os, "O_NOFOLLOW", 0)
    try:
        descriptor = os.open(path, flags)
    except OSError:
        raise RemanError("reman_pdf_changed_during_read") from None
    try:
        opened = os.fstat(descriptor)
        if not stat.S_ISREG(opened.st_mode) or not _same_file(before, opened):
            raise RemanError("reman_pdf_changed_during_read")
        chunks = []
        while True:
            chunk = os.read(descriptor, 1024 * 1024)
            if not chunk:
                break
            chunks.append(chunk)
        after = os.fstat(descriptor)
        try:
            path_after = os.lstat(path)
        except OSError:
            raise RemanError("reman_pdf_changed_during_read") from None
        if not _same_file(opened, after) or not _same_file(after, path_after):
            raise RemanError("reman_pdf_changed_during_read")
        return path, b"".join(chunks), opened.st_size
    finally:
        os.close(descriptor)

File: openclaw/reman-agentic/src/file-access.ts
import { constants } from "node:fs";
import { lstat, open, realpath, stat } from "node:fs/promises";
import { isAbsolute, relative, resolve, sep, delimiter } from "node:path";
import { RemanError } from "./client.js";

export type AllowedPdfRoot = { declared: string; canonical: string };

export function configuredPdfDirectoryValues(configured?: string[]): string[] {
  const source = configured?.length
    ? configured
    : (process.env.REMAN_AGENT_ALLOWED_PDF_DIRS ?? "").split(delimiter);
  return [...new Set(source.map((value) => value.trim()).filter(Boolean))];
}

export function hasConfiguredPdfDirectories(configured?: string[]): boolean {
  return configuredPdfDirectoryValues(configured).length > 0;
}

export async function resolveAllowedPdfRoots(configured?: string[]): Promise<AllowedPdfRoot[]> {
  const values = configuredPdfDirectoryValues(configured);
  if (!values.length) throw new RemanError("reman_pdf_roots_not_configured");
  const roots: AllowedPdfRoot[] = [];
  for (const value of values) {
    if (!isAbsolute(value)) throw new RemanError("reman_pdf_root_invalid");
    const declared = resolve(value);
    let canonical: string;
    try {
      canonical = await realpath(declared);
      if (!(await stat(canonical)).isDirectory()) throw new Error("not_directory");
    } catch {
      throw new RemanError("reman_pdf_root_invalid");
    }
    if (!roots.some((root) => root.declared === declared && root.canonical === canonical)) {
      roots.push({ declared, canonical });
    }
  }
  return roots;
}

function locateUnderRoot(rawPath: string, roots: AllowedPdfRoot[]): { path: string; root: string; parts: string[] } {
  if (!isAbsolute(rawPath) || rawPath.split(/[\\/]+/u).includes("..")) {
    throw new RemanError("reman_pdf_path_denied");
  }
  const lexical = resolve(rawPath);
  for (const allowed of roots) {
    for (const base of [allowed.declared, allowed.canonical]) {
      const child = relative(base, lexical);
      if (child && !child.startsWith(`..${sep}`) && child !== ".." && !isAbsolute(child)) {
        return { path: resolve(allowed.canonical, child), root: allowed.canonical, parts: child.split(sep) };
      }
    }
  }
  throw new RemanError("reman_pdf_path_denied");
}

function sameFile(left: Awaited<ReturnType<typeof lstat>>, right: Awaited<ReturnType<typeof lstat>>): boolean {
  return left.dev === right.dev
    && left.ino === right.ino
    && left.size === right.size
    && left.mtimeMs === right.mtimeMs
    && left.ctimeMs === right.ctimeMs;
}

export async function readAllowedPdf(options: {
  rawPath: string;
  roots: AllowedPdfRoot[];
  maxBytes?: number;
  beforeOpen?: (path: string) => Promise<void> | void;
}): Promise<{ path: string; content: Buffer; size: number }> {
  const located = locateUnderRoot(options.rawPath, options.roots);
  let current = located.root;
  let before: Awaited<ReturnType<typeof lstat>> | undefined;
  for (const [index, part] of located.parts.entries()) {
    current = resolve(current, part);
    try {
      before = await lstat(current);
    } catch {
      throw new RemanError("reman_pdf_invalid_or_missing");
    }
    if (before.isSymbolicLink()) throw new RemanError("reman_pdf_symlink_denied");
    if (index < located.parts.length - 1 && !before.isDirectory()) {
      throw new RemanError("reman_pdf_path_denied");
    }
  }
  if (!before?.isFile()) throw new RemanError("reman_pdf_not_regular");
  if (before.size <= 0 || (options.maxBytes !== undefined && before.size > options.maxBytes)) {
    throw new RemanError("reman_pdf_exceeds_grant_limit");
  }
  try {
    if ((await realpath(located.path)) !== located.path) throw new RemanError("reman_pdf_symlink_denied");
  } catch (error) {
    if (error instanceof RemanError) throw error;
    throw new RemanError("reman_pdf_invalid_or_missing");
  }
  await options.beforeOpen?.(located.path);

  const noFollow = "O_NOFOLLOW" in constants ? constants.O_NOFOLLOW : 0;
  let handle;
  try {
    handle = await open(located.path, constants.O_RDONLY | noFollow);
  } catch {
    throw new RemanError("reman_pdf_changed_during_read");
  }
  try {
    const opened = await handle.stat();
    if (!opened.isFile() || !sameFile(before, opened)) throw new RemanError("reman_pdf_changed_during_read");
    const content = await handle.readFile();
    const after = await handle.stat();
    let pathAfter;
    try {
      pathAfter = await lstat(located.path);
    } catch {
      throw new RemanError("reman_pdf_changed_during_read");
    }
    if (!sameFile(opened, after) || !sameFile(after, pathAfter)) {
      throw new RemanError("reman_pdf_changed_during_read");
    }
    return { path: located.path, content, size: opened.size };
  } finally {
    await handle.close();
  }
}
